AEPLifecycle/Sources/XDM/XDMLifecycleMetricsBuilder.swift

```diff
@@ -0,0 +1,137 @@
+        appLaunchXDMData.environment = xdmEnvironmentInfo
+        appLaunchXDMData.eventType = LifecycleConstants.XDM.EVENT_TYPE_APP_LAUNCH
+        appLaunchXDMData.timestamp = startDate
+
+        return appLaunchXDMData.asDictionary()
+    }
+
+    /// Builds the data required for the XDM Application Close event, including `XDMLifecycleApplication`
+    /// - Returns: App close event data in dictionary format
+    func buildXDMAppCloseEventData() -> [String: Any]? {
+        var appLaunchXDMData = XDMMobileLifecycleDetails()
+        appLaunchXDMData.application = xdmApplicationInfoClose
+        appLaunchXDMData.eventType = LifecycleConstants.XDM.EVENT_TYPE_APP_CLOSE
+        appLaunchXDMData.timestamp = startDate
+
+        // TODO: MOB-14370 backdate timestamp
+
+        return appLaunchXDMData.asDictionary()
+    }
+
+    /// Adds general application information as well as details related to the type of launch (install, upgrade, regular launch)
+    /// - Parameters:
+    ///   - isInstall: indicates if this is an app install
+    ///   - isUpgrade: indicates if this is an app upgrade
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addAppLaunchData(isInstall: Bool, isUpgrade: Bool) -> XDMLifecycleMetricsBuilder {
+        xdmApplicationInfoLaunch = XDMLifecycleApplication()
+        xdmApplicationInfoLaunch?.isLaunch = true
+
+        if isInstall {
+            xdmApplicationInfoLaunch?.isInstall = true
+            xdmApplicationInfoLaunch?.installDate = startDate
+        }
+
+        if isUpgrade {
+            xdmApplicationInfoLaunch?.isUpgrade = true
+        }
+
+        xdmApplicationInfoLaunch?.name = systemInfoService.getApplicationName()
+        xdmApplicationInfoLaunch?.id = systemInfoService.getApplicationBundleId()
+        xdmApplicationInfoLaunch?.version = systemInfoService.getApplicationVersion()
+
+        return self
+    }
+
+    /// Adds general application information as well as details related to the type of close event
+    /// - Parameters:
+    ///   - previousAppId: application version from the previous session when the close happened
+    ///   - previousSessionInfo: previous session information to be attached to the close event
+    /// - Returns: this `XDMLifecycleMetricsBuilder` instance
+    func addAppCloseData(previousAppId: String?, previousSessionInfo: LifecycleSessionInfo) -> XDMLifecycleMetricsBuilder {
+        xdmApplicationInfoClose = XDMLifecycleApplication()
+
+        xdmApplicationInfoClose?.version = previousAppId
+        xdmApplicationInfoClose?.isClose = true
+        xdmApplicationInfoClose?.closeType = previousSessionInfo.isCrash ? .unknown : .close
+        if let sessionLength = previousSessionInfo.sessionLength {
+            xdmApplicationInfoClose?.sessionLength = Int64(sessionLength)
+        }
+        xdmApplicationInfoClose?.name = systemInfoService.getApplicationName()
+        xdmApplicationInfoClose?.id = systemInfoService.getApplicationBundleId()
```

Comment:
This info seems redundant here as we transmit this info in the app launch event

Reply:
Aren't we doing that here? https://git.corp.adobe.com/dms-mobile/bourbon-core-java-lifecycle/blob/6dbd0b666f41af458f445b94de2bc64217c1fda7/code/lifecycle/src/main/java/com/adobe/marketing/mobile/XDMLifecycleMetricsBuilder.java#L159

Reply:
Ah.. my bad, checked the spec and they should be included.